subtemplate replacers

cmangeat opened the pull request:
based on https://github.com/geonetwork/core-geonetwork/pull/6145

fxprunayre committed:
Subtemplate / Improve indexing.

fxprunayre committed:
Indexing / ISO / Add supplementalInformation and purpose fields.

cmangeat committed:
wip, reporting high level test

fxprunayre commented on schemas/schema-core/src/main/java/org/fao/geonet/kernel/schema/subtemplate/AbstractReplacer.java:
```diff
@@ -0,0 +1,173 @@
+                .collect(Status.STATUS_COLLECTOR);
+    }
+
+    protected Status replace(Element element,
```
One thing here is that it is quite similar to existing https://github.com/geonetwork/core-geonetwork/blob/main/services/src/main/java/org/fao/geonet/api/registries/DirectoryUtils.java#L138 (not yet completely migrated to Elasticsearch). Maybe we should consider combining both approaches ?

![image](https://user-images.githubusercontent.com/1701393/140459616-c39134a3-e3b9-4571-bea9-f0dd700133d2.png)

Similarities : 
* Take a document,
* identify elements,
* check if corresponding entries are in directory,
* do something eg. here substitute as xlink

Main differences:
* DirectoryUtils is limited to one type of search and substitution at a time.
* it can work with whatever elements
* it can be played on demand like processes on uuid or a selection
* it can update the element by the directory entry AND substitute as XLink (and also create entries in the directory)
* It does exact match only
* DirectoryUtils is probably less elegant...

What do you think @cmangeat ?

cmangeat replied:
we should maybe using consider https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html rather than querying for list of identified fields so to find the matching subtemplate.

considering the possibility of automatically creating subtemplates, it depends of the workflow, here focus was avoiding proliferation of subtemplates, letting the sole "high level" admins managing contacts dictionary for example, this way offering a service to metadata publisher: the phone number are for example assured to be up to date, and so on (office, office delegate and so on).

still wip anyway.

fxprunayre commented on schemas/iso19139/src/main/java/org/fao/geonet/schema/iso19139/FormatReplacer.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    protected int queryAddExtraClauses(StringBuilder query, Element format, String lang) throws JDOMException {
+        String name = getFieldValue(format, ".//gmd:name", lang);
```
Instead of hardcoding the replacer config, should we do something more like 

```xml
  <bean id="iso19139SchemaPlugin"
        class="org.fao.geonet.schema.iso19139.ISO19139SchemaPlugin">
    <property name="replacers">
      <util:list>
        <bean id="iso191139FormatReplacer" class="Replacer">
          <property name="xpath" value=".//gmd:MD_Format/parent::*"/>
          <property name="matching">
            <util:map>
              <entry key="fname" value=".//gmd:name"/>
              <entry key="fversion" value=".//gmd:version"/>
            </util:map>
          </property>
<!--          <property name="extraClause" value="+valid:1"/> Allow to limit subtemplates to choose from -->
<!--          <property name="minWeight" value="?"/> Not really sure how this is suppose to work yet. If not set, then do exact match (search will BTW not always be exact - depends on analyzers) -->
        </bean>
        ...
      </util:list>
    </property>
```

fxprunayre commented on schemas/schema-core/src/main/java/org/fao/geonet/kernel/schema/subtemplate/SubtemplatesByLocalXLinksReplacer.java:
```diff
@@ -0,0 +1,111 @@
+    private String computeMetadataLang(Element dataXml) {
+        String lang = "eng";
+        try {
+            lang = Xml.selectElement(dataXml, ".//gmd:language/gmd:LanguageCode", namespaces).getAttributeValue("codeListValue");
```
This is a bit to ISO19139 focus, we should consider using extract-default-language.xsl. So we only search subtemplate in the main metadata language right ?

fxprunayre commented on schemas/iso19139/src/main/plugin/iso19139/index-fields/index-subtemplate.xsl:
```diff
@@ -139,6 +139,12 @@
+                          then gmd:description/gco:CharacterString
+                          else string-join(.//gco:Decimal, ', ')" />
+           store="true" index="true"/>
+    <xsl:copy-of select="gn-fn-index:add-field('ftitle', $ftitle)"/>
```
Those fields are subtemplate specific, maybe we should prefix them otherwise in tools like Kibana they will be mixed with others and user will not really get what fname is or the difference between ftitle and resourceTitle ?

fgravin committed:
feat(index): add dedicated analyzers to each language

fgravin committed:
feat(search): add an option to search only in UI language

fgravin committed:
feat(search): enable multilingual full text search

by default it searches over all languages.
you can restrict to the UI language by checking the corresponding search option.

fgravin committed:
feat(search): add multilingual for the suggestion query

fgravin committed:
doc(index): multilingual documentation

fgravin committed:
chore(git): add generated records.json in .gitignore

fgravin committed:
doc(search): explain the queryBase UI setting

fgravin committed:
feat(index): add asciifolding in english analyzer

fgravin committed:
fix(i18n): be sure the lang is translate in search options

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/core-geonetwork?pullRequest=6019) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/core-geonetwork?pullRequest=6019) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: fxprunayre<br/>:white_check_mark: cmangeat<br/>:x: fgravin<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/core-geonetwork?pullRequest=6019) it.</sub>
